src/mpi/coll/op/opequal.c

```diff
@@ -0,0 +1,178 @@
+/* Values of is_equal: 1 equal 0 not equal, init using any value other than 0.*/
+#define MPIR_EQUAL_C_CASE_INT(c_type_, _len_) {             \
+        c_type_ *a = (c_type_ *)inoutvec;                   \
+        c_type_ *b = (c_type_ *)invec;                      \
+        for (i = 0; i < _len_; ++i) {                       \
+            if (!(b[i].is_equal) || !(a[i].is_equal)){      \
+                a[i].is_equal = 0;                          \
+            }else if (a[i].value != b[i].value){            \
+                a[i].is_equal = 0;                          \
+            }else{                                          \
+                a[i].is_equal = 1;                          \
+            }                                               \
+        }                                                   \
+    }                                                   
+
+
+void MPIR_EQUAL_struct(void *invec, void *inoutvec, int *Len, MPI_Datatype * type)
+{
+    int len = *Len, data_len = 0, type_len = 0;
+    int i, j, is_equal, *ints;
+    void *invec_i_pos, *invec_i_bool_pos, *inoutvec_i_pos, *inoutvec_i_bool_pos;
+    MPI_Aint *aints, *counts, lb, extent, size;
+    MPI_Datatype *types;
+    MPIR_Datatype *typeptr;
+    MPIR_Datatype_contents *cp;
+
+    /* decode */
+    MPIR_Datatype_get_ptr(*type, typeptr);
+
+    cp = typeptr->contents;
+
+    MPIR_Datatype_access_contents(cp, &ints, &aints, &counts, &types);
+
+    MPIR_Type_get_extent_impl(*type, &lb, &extent);
+    type_len = extent - lb;
+
+    /*Compare structs*/
+    for (i = 0; i < len; ++i)       
+    {
+        invec_i_pos = invec + i*type_len;
+        invec_i_bool_pos = invec_i_pos + aints[cp->nr_aints-1];
+        inoutvec_i_pos = inoutvec + i*type_len;
+        inoutvec_i_bool_pos = inoutvec_i_pos + aints[cp->nr_aints-1];
+
+        if(!(*(int*)invec_i_bool_pos) ||
+            !(*(int*)inoutvec_i_bool_pos))
+        {/*compare values of is_equal*/
+            *(int*)inoutvec_i_bool_pos  = 0;
+        }
+        else
+        {/*compare the content of the struct*/
+            is_equal = 1;
+
+            for(j = 0; j < cp->nr_types-1; ++j)
+            {
+                MPIR_Type_size_impl(types[j], &size); 
+                data_len = ints[j + 1] * size;
+
+                if(memcmp(invec_i_pos+aints[j], inoutvec_i_pos+aints[j], data_len))
+                    is_equal = 0;
+            }
+
+            *(int*)inoutvec_i_bool_pos  = is_equal;
+        }
+    }
+}
+
+void MPIR_EQUAL(void *invec, void *inoutvec, int *Len, MPI_Datatype * type)
+{
+    int i, len = *Len;
+   
+    switch (*type) {
+            /* first the C types */
+        case MPI_2INT:
+            MPIR_EQUAL_C_CASE_INT(MPIR_2int_eqltype, len);
+            break;
+        case MPI_LONG_INT:
+            MPIR_EQUAL_C_CASE_INT(MPIR_longint_eqltype, len);
+            break;
+        case MPI_SHORT_INT:
+            MPIR_EQUAL_C_CASE_INT(MPIR_shortint_eqltype, len);
+            break;
```

Comment:
You don't need pare type cases now.

Reply:
It will be faster if using MPI_2int etc.. It takes extra steps if we assume every case is using derived datatype.
For example, to confirm if a integer equals across all processes, we can use [MPI_EQUAL MPI_2int {int, is_equal}].

Reply:
The will be no cost for constructing, commit, and free a derived datatype.

Reply:
115   mpi_errno = MPI_Type_create_struct(s_count, s_blocklengths, s_displacements, s_types, &derived_type);
116   MPI_Type_commit(&derived_type); 
...
133   MPI_Type_free(&derived_type);

Reply:
I think it's better for the user makes this decision; it's redundant if check this in` MPIR_Reduce_equal`.